### PR TITLE
Exchange for (Currency) from services

### PR DESCRIPTION
**Pull request by Swingcloud**
- What is the purpose of this Pull Request? add rescue method for possible error for calling api endpoint, and add `exchange_for` and `available_exchange`
- What is the related issue for this Pull Request? #112 
- [ ] I have added Specs
- [ ] (If implementing Market Ticker) I have verified that the `volume` attributes is correct
- [ ] (If implementing Market Ticker) I have verified that the `base` and `target` is correct

this is really slow, cuz it needs to run through all services, takes about 2 mins to do it

### CODE REVIEW

**Comment by tmlee** [comment]
@dominicwong617 Maybe you can look into this

File: lib/cryptoexchange/client.rb [hunk 1]
```diff
@@ -8,6 +8,8 @@ def initialize(ticker_ttl: 3)
     def pairs(exchange)
       pairs_classname = "Cryptoexchange::Exchanges::#{StringHelper.camelize(exchange)}::Services::Pairs"
       Object.const_get(pairs_classname).new.fetch
+    rescue HTTP::ConnectionError, HTTP::TimeoutError, JSON::ParserError
+      return { error: 'exchange service temporary unavailable' }
```
Comment (dominicwong617) [inline lib/cryptoexchange/client.rb]: @Swingcloud How about `#{exchange}'s service is temporarily unavailable.`

File: lib/cryptoexchange/client.rb [hunk 2]
```diff
@@ -24,6 +26,41 @@ def ticker(market_pair)
...
+
+    def exchange_for(currency)
+      exchanges = {
+        base: [],
+        target: [],
+        temporary_unavailable: []
+      }
+      available_exchanges.each do |exchange|
+        pairs = pairs(exchange)
+        if pairs.is_a?(Hash)
```
Comment (dominicwong617) [inline lib/cryptoexchange/client.rb]: @Swingcloud it's more like `pairs.is_a?(Hash) && pairs[:error]` ?

File: lib/cryptoexchange/client.rb [hunk 3]
```diff
@@ -24,6 +26,41 @@ def ticker(market_pair)
...
+          exchanges[:temporary_unavailable].push(exchange)
```
Comment (dominicwong617) [inline lib/cryptoexchange/client.rb]: `temporarily_unavailable` ?

File: lib/cryptoexchange/client.rb [hunk 4]
```diff
@@ -24,6 +26,41 @@ def ticker(market_pair)
...
+          next
+        else
+          pairs(exchange).each do |pair|
```
Comment (dominicwong617) [inline lib/cryptoexchange/client.rb]: @Swingcloud why do you need it call it again?

File: lib/cryptoexchange/client.rb [hunk 5]
```diff
@@ -24,6 +26,41 @@ def ticker(market_pair)
...
+            if pair.base == currency.upcase
+              exchanges[:base].push(exchange)
```
Comment (dominicwong617) [inline lib/cryptoexchange/client.rb]: @Swingcloud you need to make sure there are no duplicates

File: lib/cryptoexchange/client.rb [hunk 6]
```diff
@@ -24,6 +26,41 @@ def ticker(market_pair)
...
+              break
+            elsif pair.target == currency.upcase
+              exchanges[:target].push(exchange)
+              break
+            end
+          end
+        end
+      end
+      exchanges
```
Comment (dominicwong617) [inline lib/cryptoexchange/client.rb]: @Swingcloud actually let's just return it as one array and ignore the ones that raises a HTTP error

Comment (Swingcloud) [reply lib/cryptoexchange/client.rb]: will do

File: lib/cryptoexchange/client.rb (hunk 4 above)
Comment (Swingcloud) [reply lib/cryptoexchange/client.rb]: I call it again because some exchanges are at target, not at base

File: lib/cryptoexchange/client.rb (hunk 1 above)
Comment (Swingcloud) [reply lib/cryptoexchange/client.rb]: will do

File: lib/cryptoexchange/client.rb (hunk 4 above)
Comment (Swingcloud) [reply lib/cryptoexchange/client.rb]: my bad, I know what you mean

File: lib/cryptoexchange/client.rb [hunk 7]
```diff
@@ -24,6 +26,30 @@ def ticker(market_pair)
...
+        folder_name.split('/').last
+      end
+    end
+
+    def exchange_for(currency)
+      exchanges = []
+      available_exchanges.each do |exchange|
+        pairs = pairs(exchange)
+        next if pairs.is_a?(Hash) && !pairs[:error].empty?
+        pairs.each do |pair|
```
Comment (dominicwong617) [inline lib/cryptoexchange/client.rb]: @Swingcloud you are still going to have duplicates, can you call `uniq` in the end?

Comment (dominicwong617) [reply lib/cryptoexchange/client.rb]: also, would be great if you can document this method in README

Comment (Swingcloud) [reply lib/cryptoexchange/client.rb]: seems like if it has an equal value on base or target, that will break and go to the other exchange.

and I'll document it

File: lib/cryptoexchange/client.rb [hunk 8]
```diff
@@ -8,6 +8,8 @@ def initialize(ticker_ttl: 3)
     def pairs(exchange)
       pairs_classname = "Cryptoexchange::Exchanges::#{StringHelper.camelize(exchange)}::Services::Pairs"
       Object.const_get(pairs_classname).new.fetch
+    rescue HTTP::ConnectionError, HTTP::TimeoutError, JSON::ParserError
```
Comment (dominicwong617) [inline lib/cryptoexchange/client.rb]: @Swingcloud actually, where is `begin`?

Comment (Swingcloud) [reply lib/cryptoexchange/client.rb]: @dominicwong617  method "def" can serve as a "begin" statement
or we can put begin on it I guess, but it works the same 

https://stackoverflow.com/questions/1542672/how-does-one-use-rescue-in-ruby-without-the-begin-and-end-block

Comment (dominicwong617) [reply lib/cryptoexchange/client.rb]: @Swingcloud TIL, no worries

**Comment by dominicwong617** [comment]
Let's try @Swingcloud @tmlee

**Pull request merged**